### ISSUE TITLE
Extended variable syntax failures not caught by "wait until keywords succeeds"

### ISSUE THREAD

**Issue by laurentbristiel** [robotframework/robotframework#2024]
Extended variable syntax is [described in the user guide](http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#extended-variable-syntax)
It allows, for example, to access an element from a list with ${list[2]}.
Doc says that "If the evaluation fails because of an invalid syntax or that the queried attribute does not exist, an exception is raised and the test fails."
I expected that this failure could be caught by "Wait until keywords succeeds", but it does not seem to be the case.
Is that a bug in Robot? Or should we make the doc clearer?

Here is an example to illustrate the problem:

```robotframework
*** Settings ***
Library  Collections

*** Test Cases ***
basic variable syntax
    ${list} =  create list  first  second
    # we will actually call following statement several times during 10 seconds
    wait until keyword succeeds  10s  1s  get_third_element_from_list  ${list}

advanced variable syntax
    ${list} =  create list  first  second
    # we will run following statement only once
    # ${list[2]} failure is caught globally by the test but not locally by WUKS
    wait until keyword succeeds  10s  1s  log  ${list[2]}
      
*** Keywords ***
get_third_element_from_list
    [Arguments]  ${list}
    get from list  ${list}  2
```

**Comment by pekkaklarck** [comment]
The reason for this behavior is that all failures caused by variables not existing are considered syntax errors and Wait Until Keyword Succeeds doesn't catch them. A workaround is using Variable Should Exist that fails normally so that WUKS can catch it.

I could be easily convinced that variables not existing should not be a syntax error, but that's a backwards incompatible change and should be done in a major version. Are you @laurentbristiel interested in looking at that? It shouldn't require much more than changing variables package to use something else than `DataError` when a variable doesn't exist.

**Comment by pekkaklarck** [comment]
This is essentially a duplicate of #1869.

**Comment by laurentbristiel** [comment]
ah yes, I knew it was ringing a bell but was unable to find the other issue! Thanks for finding it. Rest of the discussion will take place there then.
